drivers/base/bootvid/i386/xbox/bootvid.c

```diff
@@ -0,0 +1,448 @@
+
+    *Back = Color;
+    *Frame = CachedPalette[Color];
+}
+
+VOID
+NTAPI
+PreserveRow(
+    _In_ ULONG CurrentTop,
+    _In_ ULONG TopDelta,
+    _In_ BOOLEAN Restore)
+{
+    PUCHAR NewPosition, OldPosition;
+
+    /* Calculate the position in memory for the row */
+    if (Restore)
+    {
+        /* Restore the row by copying back the contents saved off-screen */
+        NewPosition = BackBuffer + BB_OFFSET(0, CurrentTop);
+        OldPosition = BackBuffer + BB_OFFSET(0, SCREEN_HEIGHT);
+    }
+    else
+    {
+        /* Preserve the row by saving its contents off-screen */
+        NewPosition = BackBuffer + BB_OFFSET(0, SCREEN_HEIGHT);
+        OldPosition = BackBuffer + BB_OFFSET(0, CurrentTop);
+    }
+
+    /* Set the count and loop every pixel of backbuffer */
+    ULONG Count = TopDelta * SCREEN_WIDTH;
+
+    memcpy(NewPosition, OldPosition, Count);
+
+    if (Restore)
+    {
+        NewPosition = BackBuffer + BB_OFFSET(0, CurrentTop);
+
+        /* Set the count and loop every pixel of framebuffer */
+        for (ULONG y = 0; y < TopDelta; y++)
+        {
+            PULONG Frame = (PULONG)(FrameBufferStart + FB_OFFSET(0, CurrentTop + y));
+
+            Count = SCREEN_WIDTH;
+            while (Count--)
+            {
+                *Frame++ = CachedPalette[*NewPosition++];
+            }
+        }
+    }
+}
+
+VOID
+NTAPI
+DoScroll(
+    _In_ ULONG Scroll)
+{
+    ULONG RowSize = VidpScrollRegion[2] - VidpScrollRegion[0] + 1;
+
+    /* Calculate the position in memory for the row */
+    PUCHAR OldPosition = BackBuffer + BB_OFFSET(VidpScrollRegion[0], VidpScrollRegion[1] + Scroll);
+    PUCHAR NewPosition = BackBuffer + BB_OFFSET(VidpScrollRegion[0], VidpScrollRegion[1]);
+
+    /* Start loop */
+    for (ULONG Top = VidpScrollRegion[1]; Top <= VidpScrollRegion[3]; ++Top)
+    {
+        ULONG i;
+
+        /* Scroll the row */
+        memcpy(NewPosition, OldPosition, RowSize);
+
+        PULONG Frame = (PULONG)(FrameBufferStart + FB_OFFSET(VidpScrollRegion[0], Top));
+
+        for (i = 0; i < RowSize; ++i)
+            Frame[i] = CachedPalette[NewPosition[i]];
+
+        OldPosition += SCREEN_WIDTH;
+        NewPosition += SCREEN_WIDTH;
+    }
+}
+
+VOID
+NTAPI
+DisplayCharacter(
+    _In_ CHAR Character,
+    _In_ ULONG Left,
+    _In_ ULONG Top,
+    _In_ ULONG TextColor,
+    _In_ ULONG BackColor)
+{
+    /* Get the font and pixel pointer */
+    PUCHAR FontChar = GetFontPtr(Character);
+
+    /* Loop each pixel height */
+    for (ULONG y = 0; y < BOOTCHAR_HEIGHT; y++, Top++)
+    {
+        /* Loop each pixel width */
+        ULONG x = Left;
+
+        for (UCHAR bit = 1 << 7; bit > 0; bit >>= 1, x++)
+        {
+            /* Check if we should draw this pixel */
+            if (FontChar[y * FONT_PTR_DELTA] & bit)
+            {
+                /* We do, use the given Text Color */
+                SetPixel(x, Top, (UCHAR)TextColor);
+            }
+            else if (BackColor < BV_COLOR_NONE)
```

Comment:
```suggestion
            else if (BackColor != BV_COLOR_NONE)
```

Reply:
With the '<' the code ensures that if the user passes an invalid value greater than BV_COLOR_NONE, there will be no problems as well.